Initial SPI driver support for TI OMAP and K3 devices

natto1784 opened the pull request:
This patch adds the initial support for the OMAP Multi-Channel SPI. Some things should be noted however:

#### DMA xfers are not supported yet

Only PIO is supported as of now.

#### Multi-Channel controller is not supported yet

Only Single-Channel controller mode is supported, this means that the controller can xfer messages with one slave at a time.

Passes all loopback tests on AM64x/AM243x

tags: TI, OMAP, K3, SPI

Commit:
drivers: spi: introduce TI omap_mcspi

This patch adds the initial support for the OMAP Multi-Channel SPI. Some
things should be noted however:

- DMA xfers are not supported yet. Only PIO is supported as of now.
- Multi-Channel controller is not supported yet. Only single-channel
  controller mode is supported, this means that the controller can xfer
  messages with one slave at a time.

Signed-off-by: Amneesh Singh <a-singh7@ti.com>

gramsay0 reviewed (commented):
Please update the first line of the commit message e.g:

> drivers: spi: Introduce TI spi_omap_mcspi

natto1784 commented on dts/arm/ti/am64x_m4.dtsi:
ah lol an artifact from my other PR 😶

natto1784 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,563 @@
+	volatile uint32_t *tx_reg = &regs->CHAN[chan].TX;
+	volatile uint32_t *rx_reg = &regs->CHAN[chan].RX;
+
+#define WAIT_FOR_STAT(txrx)                                                                        \
```
I was expecting this to be brought up. The thing is that if you notice where these macros are used, you will find that the `tx` and `rx` variables have different types, hence, I cannot wrap it into a function with the same function signature for all 3 cases. So it's either that I use a macro, or I simply repeat it (this is what the linux driver does).

gramsay0 replied:
You shouldn't need to repeat the whole section of code, just add a helper function to do the generic bit:
```c
static ALWAYS_INLINE uint32_t tx_frame_get(const struct spi_context *ctx, uint8_t dfs)
{
	switch (dfs) {
	case 1:
		return UNALIGNED_GET((uint8_t *)ctx->tx_buf);
	case 2:
		return UNALIGNED_GET((uint16_t *)ctx->tx_buf);
	case 4:
		return UNALIGNED_GET((uint32_t *)ctx->tx_buf);
	}
}
```
e.g:
```c
*tx_reg = tx_frame_get(ctx, data->dfs);
ctx->tx_buf += dfs;
```

natto1784 replied:
Right, this is what I had as well iirc, but decided against it since upon checking assembly, the conditional won't be hoisted outside the do-while-loop. But if it's not too much overhead, we can have this.

natto1784 replied:
Okay did it the way you suggested, can you see if it looks okay now. I guess we can live with this much overhead.

gramsay0 reviewed (commented):
Most SPI drivers also call `spi_context_cs_control()` to enable chip select GPIO emulation

natto1784 commented:
added, also added `spi_cs_is_gpio` instead of manual check

Ayush1325 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,600 @@
+	volatile uint32_t *sysstatus = &regs->SYSSTATUS;
+	int ret;
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
```
`DEV_REGS` should not be used before `DEVICE_MMIO_MAP` is called. This will fail in RAM MMIO.
See https://github.com/zephyrproject-rtos/zephyr/pull/90161 and https://github.com/zephyrproject-rtos/zephyr/pull/89788

natto1784 replied:
Oh thanks, completely missed this

tbursztyka reviewed (commented):
tiny comments,  lgtm otherwise

natto1784 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,600 @@
+	DEVICE_MMIO_ROM;
+	const struct pinctrl_dev_config *pinctrl;
+	uint8_t num_cs;
+	uint32_t clock_frequency;
```
done, should have sorted by length earlier!

natto1784 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,600 @@
+struct omap_mcspi_data {
+	DEVICE_MMIO_RAM;
+	struct spi_context ctx;
+	uint8_t dfs; /* data frame size - word length in bytes */
```
done

gramsay0 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,600 @@
+		clkg = 0;
+	} else {
+		divisor = (ref_hz + speed_hz - 1) / speed_hz;
+		speed_hz = ref_hz / divisor;
```
The sonarqube bot noticed the value assigned here is unused

gramsay0 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,599 @@
+
+	speed_hz = MIN(speed_hz, ref_hz);
+	if (speed_hz < (ref_hz / OMAP_MCSPI_MAX_DIVIDER)) {
+		speed_hz = ref_hz >> clkd;
```
`>> clkd` appears unnecessary, it always shifts by zero bits.

Also, why is `speed_hz` getting set to `ref_hz`?
It seems like it ignores the requested `speed_hz` and always result in a divisor of 1

natto1784 replied:
lines 145 and 146 need to be swapped. that will calculate the value of `clkd` using `speed_hz` and the max frequency (`ref_hz`). thanks for pointing this out.

dnltz reviewed (dismissed):
Looks good to me. Can you please take a look at the those issues https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=87970&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true ?

You should also add this device to at least one device-tree or create a test case. Otherwise, github actions won't compile this driver.

kartben commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,596 @@
+
+	while (!(*reg & bit)) {
+		/* 100ms timeout */
+		if (k_uptime_get() - timeout < 100) {
```
The condition is backwards?..
Also, using k_uptime_delta() might be nice

natto1784 replied:
The condition is correct, we loop until the bit is set, I shall make it clearer

gramsay0 replied:
@kartben is correct, it should timeout if the elapsed time is _greater than_ 100

kartben replied:
Not sure why I dismissed my review, sorry, I thought you had fixed it but I might have mixed up PRs. It is a bug that needs to be fixed, yes :)

vaishnavachath commented:
> > Looks good to me. Can you please take a look at the those issues
> 
> Those are all "if-without-else" issues, I can add an else clause if those need to be addressed urgently.
> 
> > You should also add this device to at least one device-tree or create a test case.
> 
> I have only tested this on AM243x EVM so that is where I will be adding this and it is not merged yet unfortunately.

@natto1784 can the DT changes also be added now that the platform support is merged?

vaishnavachath reviewed (changes requested):
please add the SPI nodes for devices tested.

Ayush1325 commented:
> please add the SPI nodes for devices tested.

I would like to add that please use `{main/mcu}_spi*` if you are adding new nodes.

natto1784 commented:
Some minor changes

1. Rewrote clk configuration to be more clear and apparent
2. While reading RX register, we only read the configured number of bits (e.g. 7 or 9)
3. Updated calls for `spi_context_tx/rx_len_left`

Tomorrow, will add this to the am243x along with the loopback test overlay

gramsay0 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,597 @@
+
+	while ((*reg & bit) == 0) {
+		/* 100ms timeout */
+		if (k_uptime_delta(&timeout) < 100) {
```
`k_uptime_delta` updates `timeout` to the current time on each call
https://docs.zephyrproject.org/apidoc/latest/group__clock__apis.html#gad748b2fe83b36884dc087b4af367de80
`sys_timepoint_expired` might be a better option.
(and as mentioned elsewhere `<` should be `>`)

natto1784 replied:
Makes sense, will change

gramsay0 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,597 @@
+		fRatio++;
+	}
+
+	if ((fRatio & (fRatio - 1)) != 0) {
```
Thanks for the explanation!

Ok a high resolution 12bit range (granularity=1) and a low resolution 15bit range (granularity=power of 2).

Why use the low resolution divide for values that the high resolution can accommodate (just because they happen to be an exact power of two)?

Would it make sense to use the low resolution divide only if it does not fit in the high resolution (12bit) range?

I think there should be some bounds checking too.
i.e. if you request 10kHz at 48MHz reference clock the 12bit range would overflow

natto1784 replied:
Aye, we should add some bound checking. The upper limit for the divider is 4096, so I suppose we can use that.

gramsay0 replied:
I think you overlooked what I was saying...

4096 is the max divide if using the extended 12bit (granularity=1) divider.
The power of two divider is 4bits (0->15) that represent 2^**0** -> 2^**15** (i.e. up to 32768)

Why use the low resolution divide for values that the high resolution can accommodate (just because they happen to be an exact power of two)?

Would it make sense to use the low resolution divide only if it does not fit in the high resolution (12bit) range?

![image](https://github.com/user-attachments/assets/d07c55d8-c983-42cf-8634-2fc9674cfdd3)
![image](https://github.com/user-attachments/assets/32a32751-ad0d-468d-8099-1d86f2280f77)
![image](https://github.com/user-attachments/assets/1847b7dc-d283-4215-88a2-b17e9ad62329)

natto1784 replied:
Well, yes we can do this and it shouldn't cause any difference. I just felt that we ought to keep things default as much as possible, hence I avoided using CLKG if possible.

natto1784 commented:
There are quite some changes with the new version:

1. Remove `OMAP_MCSPI_MAX_FREQUENCY` in favor of getting the controller frequency from DT only https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L655
2. Configure and use FIFO.
3. Return `-ENOTSUP` for `SPI_HOLD_ON_CS` https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L200-L202
4. Use 12-bit resolution (granularity of 1) for clock dividers less than 4096, and the 15-bit resolution (for powers of 2 only) otherwise. https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L151-L166
5. Do not use timeout when waiting for register status, instead check the register for only a 100 times without waiting.  https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L305-L316
6. Rewrite the `omap_mcspi_transceive_pio` to highlight TX only, RX only and TX/RX chunks separately and make sure to fill/empty FIFO whenever possible. https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L378-L450
7. In `omap_mcspi_transceive_one`, configure RX and TX separately for more code clarity. https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L477-L503
8. Don't do anything if both RX and TX buffers are NULL and move on to the next buffers in the set. https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L468-L470
9. Use `word_mask` to get the intended amount of bits from the word (for example, getting a word of 5 bits using `0b11111` on an 8 bit integer)  https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L367
10. Write `count` to `WCNT` so that we can keep track of the words transferred using FIFO. https://github.com/zephyrproject-rtos/zephyr/blob/5572b47715f101beef6ff2deb812535828c5f2cd/drivers/spi/spi_omap_mcspi.c#L510

Commit:
am243x_evm/am2434/r5f0_0: add SPI support

Add OMAP multi-channel SPI node to the device tree and add overlay for the
SPI loopback test.

Signed-off-by: Amneesh Singh <a-singh7@ti.com>

gramsay0 commented:
> This is done to avoid the massive overhead `k_uptime_get` was causing.

Out of interest, how long does a call to `k_uptime_get` take?
It should be pretty fast, maybe there is a bug if it is really slow

gramsay0 commented on boards/ti/am243x_evm/am243x_evm_am2434_r5f0_0-pinctrl.dtsi:
```diff
@@ -24,6 +24,22 @@
+	};
+
+	spi0_clk: spi0_clk_default {
+		pinmux = <K3_PINMUX(0x210, PIN_INPUT, MUX_MODE_0)>;
```
Should clock and MOSI be `PIN_OUTPUT`?

gramsay0 reviewed (changes requested):
I ran the SPI loopback tests on AM6234 Cortex-A53 using MCU_MCSPI0.
Mostly worked really well! Just a couple minor things to tidy up

I also noticed this test failed:
```
START - test_spi_complete_multiple_timed
Transfer took 139 us vs theoretical minimum 8 us
Latency measurement: 131 us
```
100us seems fast to configure the SPI and run the test.
The test SPI frequency for this device is 50MHz, the next highest I spotted was 16MHz.
Maybe the test should allow some fixed processing overhead time?

natto1784 commented:
> > This is done to avoid the massive overhead `k_uptime_get` was causing.
> 
> Out of interest, how long does a call to `k_uptime_get` take? It should be pretty fast, maybe there is a bug if it is really slow

Roughly around 5x the current time, maybe something to do with dmtimer? I will check

natto1784 commented on drivers/spi/spi_omap_mcspi.c:
```diff
@@ -0,0 +1,679 @@
+cleanup:
+	spi_context_cs_control(ctx, false);
+
+	if (!(ctx->config->operation & SPI_LOCK_ON)) {
```
will replace with just `config`, thanks

natto1784 commented:
added a `ti,d1-in-d0-out` option to allow user to configure d1/d0 as MISO/MOSI respectively instead of the default d0/d1

gramsay0 commented:
> The test passes as long as the time elapsed is less than `theoretical_minimum * ideal_mult`. This `ideal_mult` is by default 8 but I have set it as 12 for `am243x_evm_am2434_r5f0_0` as it was taking around `75us` at `50MHz` for me. We can increase this number from 12 to say 25 if necessary.

I think the issue is more to do with the high SPI clock frequency. It's suggesting the theoretical minimum as 8us, but there's no way to do all that config and the transfer anywhere near that fast (even with 200MHz CPU clock).

I think the test would be better to allow some fixed overhead (say 100us) + theoretical_minimum * multiplier. Or alternatively, transmit a very long byte stream to negate the overheads.

Outside the scope of this PR though

gramsay0 commented on dts/bindings/spi/ti,omap-mcspi.yaml:
```diff
@@ -0,0 +1,23 @@
+    description: |
+      Number of chipselects supported by controller
+
+  ti,d1-in-d0-out:
```
This naming should be `d1-miso-d0-mosi` as in/out depends on master/slave

natto1784 commented:
fixed compliance
